### ISSUE TITLE
Add configurable vehicle exit side?

### ISSUE THREAD

**Issue by kenny-evitt** [MrFastZombie/ConfigurableVehicles#6]
Thanks for the mod!

I would really really like to be able to configure the engineer to exit trains on the _right_ side.

I have no idea how hard this would be to add.

I could _possibly_ help some, e.g. API research.

**Comment by kenny-evitt** [comment]
Possible outline:

- Register a callback for [`on_player_driving_changed_state`](https://lua-api.factorio.com/stable/events.html#on_player_driving_changed_state)
- When the engineer exits a vehicle, [`teleport`](https://lua-api.factorio.com/stable/classes/LuaControl.html#teleport) them to 'the right (hand) side' of the vehicle using the vehicle's [`direction`](https://lua-api.factorio.com/stable/classes/LuaEntity.html#direction) (and some vector math).
- Probably should also use [`can_place_entity`](https://lua-api.factorio.com/stable/classes/LuaSurface.html#can_place_entity) too (before teleporting the engineer)

**Comment by MrFastZombie** [comment]
Good idea, and thank you for giving some possible API usage for it. I'm working on a new mod so I might be slow, but I'll look into this and see if I can do it.

**Comment by MrFastZombie** [comment]
Apologies for taking so long for this, but I've come back to it and implemented it in 1.2.0. 

Not sure if the way I detect if a player leaves is robust for multiplayer or not. If you do multiplayer, lemme know if it does anything weird if you end up using it.

 The hardest part of implementing this was doing the algebra to get the point on the opposite side of the vehicle, but I think I've got it. I ended up using surface.find_non_colliding_position() to place the player correctly, so you should not get placed inside of objects.

It'll work on modded vehicles too, but for those you need to put the vehicle prototype name in a text box. The setting's tooltip says how you can get the vehicle prototype name. 

The relevant settings will be in the per user settings section, and will not need a reload to apply.

This update will be for 2.1, but I could backport it to 2.0 if desired. I was intending to, but the modding toolkit update isn't backwards compatible so it's kind of annoying to switch back and I figure most users will be moving to 2.1 when it goes stable anyways.

**Comment by kenny-evitt** [comment]
No worries on the time and thanks for implementing this! 🙂
